### PR TITLE
Adding support for Apache CloudStack as a cloudprovider

### PR DESCRIPTION
**Pull request by davidjumani**
Adding support for [Apache CloudStack](https://github.com/apache/cloudstack) as a cloudprovider for the cluster-autoscaler
Link to the CloudStack PR for this - https://github.com/apache/cloudstack/pull/4329

### CODE REVIEW

**Comment by MaciekPytel** [comment]
There are some formal requirements for adding new provider:
1. We can't accept any contribution unless you sign the cla.
2. We require any new cloudprovider implementation to have an OWNERS file with the list of its maintainers.

**Comment by davidjumani** [comment]
Sure thing @MaciekPytel I'll add the owners and sign the CLA soon

**Comment by davidjumani** [comment]
@MaciekPytel Made the changes and signed the CLA. Please let me know if any improvements can be made

**Comment by davidjumani** [comment]
### Testing :
Increased the load on the cluster and the autoscaler kicked in increasing the cluster size. Verified in the autoscaler logs as well as the CloudStack scaling API was hit. Ditto for scale down.


Scale Up
```
I1023 10:24:21.321734       1 static_autoscaler.go:229] Starting main loop
I1023 10:24:21.321939       1 client.go:169] NewAPIRequest API request URL:http://10.10.2.197:8080/client/api?apiKey=***&command=listKubernetesClusters&id=43625790-9620-45c1-95e0-c6316d0117e0&response=json&signature=***
I1023 10:24:21.342735       1 client.go:175] NewAPIRequest response status code:200
I1023 10:24:21.343863       1 cloudstack_manager.go:88] Got cluster : &{43625790-9620-45c1-95e0-c6316d0117e0 cluster-2 2 3 1 1 [0xc0006f69f0 0xc0006f6a20]}
I1023 10:24:21.346996       1 filter_out_schedulable.go:65] Filtering out schedulables
I1023 10:24:21.347089       1 filter_out_schedulable.go:132] Filtered out 0 pods using hints
I1023 10:24:21.347529       1 filter_out_schedulable.go:170] 10 pods were kept as unschedulable based on caching
I1023 10:24:21.347535       1 filter_out_schedulable.go:171] 0 pods marked as unschedulable can be scheduled.
I1023 10:24:21.347541       1 filter_out_schedulable.go:82] No schedulable pods
I1023 10:24:21.347573       1 klogx.go:86] Pod default/php-apache-79544c9bd9-qpxdf is unschedulable
I1023 10:24:21.347577       1 klogx.go:86] Pod default/php-apache-79544c9bd9-pl528 is unschedulable
I1023 10:24:21.347580       1 klogx.go:86] Pod default/php-apache-79544c9bd9-fltfz is unschedulable
I1023 10:24:21.347583       1 klogx.go:86] Pod default/php-apache-79544c9bd9-vzmrl is unschedulable
I1023 10:24:21.347585       1 klogx.go:86] Pod default/php-apache-79544c9bd9-x4qqk is unschedulable
I1023 10:24:21.347588       1 klogx.go:86] Pod default/php-apache-79544c9bd9-fgvlg is unschedulable
I1023 10:24:21.347591       1 klogx.go:86] Pod default/php-apache-79544c9bd9-5rphq is unschedulable
I1023 10:24:21.347593       1 klogx.go:86] Pod default/php-apache-79544c9bd9-bc272 is unschedulable
I1023 10:24:21.347596       1 klogx.go:86] Pod default/php-apache-79544c9bd9-44svc is unschedulable
I1023 10:24:21.347598       1 klogx.go:86] Pod default/php-apache-79544c9bd9-nlzhr is unschedulable
I1023 10:24:21.347601       1 klogx.go:86] Pod default/php-apache-79544c9bd9-fctw6 is unschedulable
I1023 10:24:21.347616       1 scale_up.go:364] Upcoming 0 nodes
I1023 10:24:21.348158       1 scale_up.go:456] Best option to resize: 43625790-9620-45c1-95e0-c6316d0117e0
I1023 10:24:21.348166       1 scale_up.go:460] Estimated 2 nodes needed in 43625790-9620-45c1-95e0-c6316d0117e0
I1023 10:24:21.348176       1 scale_up.go:574] Final scale-up plan: [{43625790-9620-45c1-95e0-c6316d0117e0 1->3 (max: 3)}]
I1023 10:24:21.348187       1 scale_up.go:663] Scale-up: setting group 43625790-9620-45c1-95e0-c6316d0117e0 size to 3
I1023 10:24:21.348240       1 cloudstack_node_group.go:56] Increase Cluster : 43625790-9620-45c1-95e0-c6316d0117e0 by 2
I1023 10:24:21.348301       1 client.go:169] NewAPIRequest API request URL:http://10.10.2.197:8080/client/api?apiKey=***&command=scaleKubernetesCluster&id=43625790-9620-45c1-95e0-c6316d0117e0&response=json&size=3&signature=***
I1023 10:24:21.348650       1 event_sink_logging_wrapper.go:48] Event(v1.ObjectReference{Kind:"ConfigMap", Namespace:"kube-system", Name:"cluster-autoscaler-status", UID:"001cc651-1508-403d-9bd3-691f96d9161b", APIVersion:"v1", ResourceVersion:"20692", FieldPath:""}): type: 'Normal' reason: 'ScaledUpGroup' Scale-up: setting group 43625790-9620-45c1-95e0-c6316d0117e0 size to 3
I1023 10:24:21.385772       1 client.go:175] NewAPIRequest response status code:200
I1023 10:24:31.386255       1 client.go:169] NewAPIRequest API request URL:http://10.10.2.197:8080/client/api?apiKey=***&command=queryAsyncJobResult&jobid=3ecfbb2a-49a3-48e7-bafe-2e160039cd75&response=json&signature=***
I1023 10:24:31.404068       1 client.go:175] NewAPIRequest response status code:200
I1023 10:24:31.404271       1 client.go:110] Still waiting for job 3ecfbb2a-49a3-48e7-bafe-2e160039cd75 to complete

................. 3 minutes later .................

I1023 10:27:51.386212       1 client.go:169] NewAPIRequest API request URL:http://10.10.2.197:8080/client/api?apiKey=***&command=queryAsyncJobResult&jobid=3ecfbb2a-49a3-48e7-bafe-2e160039cd75&response=json&signature=***
I1023 10:27:51.405414       1 client.go:175] NewAPIRequest response status code:200
I1023 10:27:51.407033       1 cloudstack_manager.go:103] Scaled up cluster : &{43625790-9620-45c1-95e0-c6316d0117e0 cluster-2 2 3 3 1 [0xc000bf78f0 0xc000bf7980 0xc000bf79b0 0xc000bf7a40]}
I1023 10:27:51.407087       1 metrics.go:300] Function scaleUp took 3m30.059536612s to complete
I1023 10:27:51.407106       1 eventing_scale_up_processor.go:47] Skipping event processing for unschedulable pods since there is a ScaleUp attempt this loop
I1023 10:27:51.418253       1 metrics.go:300] Function main took 3m30.096536556s to complete
I1023 10:27:58.667845       1 reflector.go:246] Listing and watching *v1.CSINode from k8s.io/client-go/informers/factory.go:134
E1023 10:27:58.671861       1 reflector.go:129] k8s.io/client-go/informers/factory.go:134: Failed to watch *v1.CSINode: failed to list *v1.CSINode: the server could not find the requested resource
```

Scale Down:
```
I1023 11:25:27.780410       1 static_autoscaler.go:229] Starting main loop
I1023 11:25:27.780637       1 client.go:169] NewAPIRequest API request URL:http://10.10.2.197:8080/client/api?apiKey=***&command=listKubernetesClusters&id=43625790-9620-45c1-95e0-c6316d0117e0&response=json&signature=***
I1023 11:25:27.805500       1 client.go:175] NewAPIRequest response status code:200
I1023 11:25:27.806148       1 cloudstack_manager.go:88] Got cluster : &{43625790-9620-45c1-95e0-c6316d0117e0 cluster-2 2 3 3 1 [0xc000a68450 0xc000a68480 0xc000a684b0 0xc000a684e0]}
I1023 11:25:27.806713       1 filter_out_schedulable.go:65] Filtering out schedulables
I1023 11:25:27.806721       1 filter_out_schedulable.go:132] Filtered out 0 pods using hints
I1023 11:25:27.806729       1 filter_out_schedulable.go:170] 0 pods were kept as unschedulable based on caching
I1023 11:25:27.806735       1 filter_out_schedulable.go:171] 0 pods marked as unschedulable can be scheduled.
I1023 11:25:27.806744       1 filter_out_schedulable.go:82] No schedulable pods
I1023 11:25:27.806835       1 static_autoscaler.go:402] No unschedulable pods
I1023 11:25:27.806849       1 static_autoscaler.go:449] Calculating unneeded nodes
I1023 11:25:27.806869       1 scale_down.go:421] Node cluster-2-node-175543dd419 - memory utilization 0.264053
I1023 11:25:27.806924       1 scale_down.go:421] Node cluster-2-node-17554fca89b - memory utilization 0.105621
I1023 11:25:27.806961       1 scale_down.go:421] Node cluster-2-node-17554fcb832 - memory utilization 0.105621
I1023 11:25:27.806974       1 scale_down.go:487] Scale-down calculation: ignoring 1 nodes unremovable in the last 5m0s
I1023 11:25:27.807045       1 cluster.go:147] Fast evaluation: cluster-2-node-17554fcb832 for removal
I1023 11:25:27.807160       1 cluster.go:184] Fast evaluation: node cluster-2-node-17554fcb832 may be removed
I1023 11:25:27.807198       1 scale_down.go:543] Finding additional 1 candidates for scale down.
I1023 11:25:27.807205       1 cluster.go:147] Fast evaluation: cluster-2-node-175543dd419 for removal
I1023 11:25:27.807213       1 cluster.go:167] Fast evaluation: node cluster-2-node-175543dd419 cannot be removed: non-daemonset, non-mirrored, non-pdb-assigned kube-system pod present: metrics-server-7c85689fb8-9ttzq
I1023 11:25:27.807295       1 scale_down.go:591] 1 nodes found to be unremovable in simulation, will re-check them at 2020-10-23 11:30:27.780375364 +0000 UTC m=+7659.668657142
I1023 11:25:27.807370       1 static_autoscaler.go:492] cluster-2-node-17554fcb832 is unneeded since 2020-10-23 11:18:15.84935054 +0000 UTC m=+6927.737632318 duration 7m11.931024824s
I1023 11:25:27.807412       1 static_autoscaler.go:492] cluster-2-node-17554fca89b is unneeded since 2020-10-23 11:15:25.092758539 +0000 UTC m=+6756.981040397 duration 10m2.687616745s
I1023 11:25:27.807489       1 static_autoscaler.go:503] Scale down status: unneededOnly=false lastScaleUpTime=2020-10-23 10:24:21.321705779 +0000 UTC m=+3693.209987547 lastScaleDownDeleteTime=2020-10-23 09:23:07.21202912 +0000 UTC m=+19.100310878 lastScaleDownFailTime=2020-10-23 09:23:07.212029229 +0000 UTC m=+19.100310988 scaleDownForbidden=false isDeleteInProgress=false scaleDownInCooldown=false
I1023 11:25:27.807510       1 static_autoscaler.go:516] Starting scale down
I1023 11:25:27.807541       1 scale_down.go:790] cluster-2-node-17554fcb832 was unneeded for 7m11.931024824s
I1023 11:25:27.807618       1 scale_down.go:790] cluster-2-node-17554fca89b was unneeded for 10m2.687616745s
I1023 11:25:27.807636       1 scale_down.go:1053] Scale-down: removing empty node cluster-2-node-17554fca89b
I1023 11:25:27.807876       1 event_sink_logging_wrapper.go:48] Event(v1.ObjectReference{Kind:"ConfigMap", Namespace:"kube-system", Name:"cluster-autoscaler-status", UID:"001cc651-1508-403d-9bd3-691f96d9161b", APIVersion:"v1", ResourceVersion:"29087", FieldPath:""}): type: 'Normal' reason: 'ScaleDownEmpty' Scale-down: removing empty node cluster-2-node-17554fca89b
I1023 11:25:27.824774       1 delete.go:103] Successfully added ToBeDeletedTaint on node cluster-2-node-17554fca89b
I1023 11:25:27.825138       1 client.go:169] NewAPIRequest API request URL:http://10.10.2.197:8080/client/api?apiKey=***&command=scaleKubernetesCluster&id=43625790-9620-45c1-95e0-c6316d0117e0&nodeids=ce790e94-b632-473e-ad26-01683b64b716&response=json&signature=***
I1023 11:25:27.859971       1 client.go:175] NewAPIRequest response status code:200
I1023 11:25:37.838493       1 static_autoscaler.go:229] Starting main loop
I1023 11:25:37.860350       1 client.go:169] NewAPIRequest API request URL:http://10.10.2.197:8080/client/api?apiKey=***&command=queryAsyncJobResult&jobid=ebdb772f-3adb-484a-b46d-66dcd1580d45&response=json&signature=***
I1023 11:25:37.877933       1 client.go:175] NewAPIRequest response status code:200
I1023 11:25:37.878044       1 client.go:110] Still waiting for job ebdb772f-3adb-484a-b46d-66dcd1580d45 to complete
I1023 11:25:38.272338       1 reflector.go:518] k8s.io/client-go/informers/factory.go:134: Watch close - *v1.StatefulSet total 0 items received
I1023 11:25:44.176987       1 reflector.go:518] k8s.io/client-go/informers/factory.go:134: Watch close - *v1beta1.PodDisruptionBudget total 0 items received
I1023 11:25:47.861083       1 client.go:169] NewAPIRequest API request URL:http://10.10.2.197:8080/client/api?apiKey=***&command=queryAsyncJobResult&jobid=ebdb772f-3adb-484a-b46d-66dcd1580d45&response=json&signature=***
I1023 11:25:47.883291       1 client.go:175] NewAPIRequest response status code:200
I1023 11:25:47.885546       1 cloudstack_manager.go:117] Scaled down cluster : &{43625790-9620-45c1-95e0-c6316d0117e0 cluster-2 2 3 2 1 [0xc0006d79b0 0xc0006d79e0 0xc0006d7a10]}
I1023 11:25:47.886065       1 client.go:169] NewAPIRequest API request URL:http://10.10.2.197:8080/client/api?apiKey=***&command=listKubernetesClusters&id=43625790-9620-45c1-95e0-c6316d0117e0&response=json&signature=***
I1023 11:25:47.886551       1 event_sink_logging_wrapper.go:48] Event(v1.ObjectReference{Kind:"Node", Namespace:"", Name:"cluster-2-node-17554fca89b", UID:"a6b0668b-0c54-4aeb-91a0-70878c1a37a3", APIVersion:"v1", ResourceVersion:"28992", FieldPath:""}): type: 'Normal' reason: 'ScaleDown' node removed by cluster autoscaler
I1023 11:25:47.914672       1 client.go:175] NewAPIRequest response status code:200
I1023 11:25:47.916126       1 cloudstack_manager.go:88] Got cluster : &{43625790-9620-45c1-95e0-c6316d0117e0 cluster-2 2 3 2 1 [0xc0005a2db0 0xc0005a2de0 0xc0005a2e10]}
I1023 11:25:47.916422       1 taints.go:77] Removing autoscaler soft taint when creating template from node
I1023 11:25:47.916966       1 metrics.go:300] Function updateClusterState took 10.078192859s to complete
```

File: cluster-autoscaler/cloudprovider/builder/builder_cloudstack.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
+// +build cloudstack
+
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (rhtyd) [inline cluster-autoscaler/cloudprovider/builder/builder_cloudstack.go:4]: Copyright 2020?

**Review by rhtyd — APPROVED** [review]
LGTM (I'm one of the PMC/committers from the apache/cloudstack project)

@davidjumani please also provide link to your apache/cloudstack PR as well as smoketests results against CloudStack master.

**Comment by davidjumani** [comment]
Link to the CloudStack PR for this - https://github.com/apache/cloudstack/pull/4329

**Comment by rhtyd** [comment]
/assign @mwielgus
/assign @aleksandra-malinowska
/assign @vivekbagade
/assign @losipiuk

@mwielgus @vivekbagade @losipiuk @alexanderkjeldaas can you help review? Thanks.

**Review by mwielgus — APPROVED** [review]
/lgtm
/approve

**Pull request merged**